src/Microsoft.Identity.Web.TokenCache/MsalAbstractTokenCacheProvider.cs

```diff
@@ -98,7 +112,23 @@ private async Task OnBeforeAccessAsync(TokenCacheNotificationArgs args)
             {
                 byte[] tokenCacheBytes = await ReadCacheBytesAsync(args.SuggestedCacheKey, CreateHintsFromArgs(args)).ConfigureAwait(false);
 
-                args.TokenCache.DeserializeMsalV3(UnprotectBytes(tokenCacheBytes), shouldClearExistingCache: true);
+                try
+                {
+                    args.TokenCache.DeserializeMsalV3(UnprotectBytes(tokenCacheBytes), shouldClearExistingCache: true);
+                }
+                catch (Exception ex)
+                {
+                    // clears the cache
+                    args.TokenCache.DeserializeMsalV3(null, shouldClearExistingCache: true);
```

Comment:
What if you have several instances of a web app or web API, and they all share the cache and sharing the decryption keys. One has misconfigured keys and clears the cache, but this will impact the others using the cache. Is this okay?

Reply:
Yes, that could happen. If you have a lot of machines which use their own keys, then each machine will retrieve token from AAD and keep overwriting the cache entry. Thinking further, Line 122 doesn't even matter much, as even without explicitly clearing the internal cache, it will still be empty, which I think I saw in testing.
Example scenario:
1. Id.Web instantiates MSAL, internal cache is empty.
2. This callback is called to get the cache entry from external cache, but fails to decrypt it.
3. Exception is caught, cache is still empty. (Clearing the cache here doesn't matter).
4. MSAL doesn't find token in internal cache. Retrieves one from AAD.
5. New token is serialized with the same cache key, encrypted, and overwriting the existing cache entry.

I think @bgavrilMS suggested adding something like an encryption key identifier to the cache key.

Reply:
This is a great question. On desktop apps, we do indeed delete the cache if we cannot decrypt it. This is done in order to not crash the app. Works great for long lived apps, but it **silently hides errors.**

So any solution where we hide exception is a compromise between: 

- do not crash the app 
- do not hide the error because it can lead to unexpected behavior 

For L2 cache `id.web.TokenCache` already made a choice to not crash the app. This error occurs lower down the stack though, i.e. we are able to read data from L2, but we cannot decrypt this data. 

My 2c on this, after reading a bunch about DataProtection API (see my next comment for details), is to not hide this error. We should however improve the devex, by re-throwing a more meaningful exception.

Thoughts @jennyf19 and @pmaytak ? CC @henrik-me and @rymeskar as well for an opinion.

Reply:
TLDR what the customer should try to do when configuring DataProtection API:

```csharp

.UnprotectKeysWithAnyCertificate(
        new X509Certificate2("certificate_1.pfx", builder.Configuration["CertificatePassword_1"]),
        new X509Certificate2("certificate_2.pfx", builder.Configuration["CertificatePassword_2"]));

```

Reply:
Made the changes here: 779904c45dbedd08585f3949e2279b6d3de1532d

> Maybe add to the [retry mechanism](https://github.com/AzureAD/microsoft-identity-web/blob/bd48335ced3fdff85d83cf4e19f8c5fcc04cc982/src/Microsoft.Identity.Web.TokenCache/Distributed/MsalDistributedTokenCacheAdapter.cs#L321)?

The retry happens before the call to deserialize. We don't know if the deserialization failed at that time.